drivers/memc/memc_stm32_sdram.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2020 Teslabs Engineering S.L.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_fmc_sdram
+
+#include <device.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(memc_stm32_sdram, CONFIG_MEMC_LOG_LEVEL);
+
+/** SDRAM controller register offset. */
+#define SDRAM_OFFSET 0x140U
+
+/** FMC SDRAM controller bank configuration fields. */
+struct memc_stm32_sdram_bank_config {
+	FMC_SDRAM_InitTypeDef init;
+	FMC_SDRAM_TimingTypeDef timing;
+};
+
+/** FMC SDRAM controller configuration fields. */
+struct memc_stm32_sdram_config {
+	FMC_SDRAM_TypeDef *sdram;
+	uint32_t power_up_delay;
+	uint8_t num_auto_refresh;
+	uint16_t mode_register;
+	uint16_t refresh_rate;
+	const struct memc_stm32_sdram_bank_config *banks;
+	size_t banks_len;
+};
+
+static int memc_stm32_sdram_init(const struct device *dev)
+{
+	const struct memc_stm32_sdram_config *config = dev->config;
+
+	SDRAM_HandleTypeDef sdram = { 0 };
+	FMC_SDRAM_CommandTypeDef sdram_cmd = { 0 };
+
+	sdram.Instance = config->sdram;
+
+	for (size_t i = 0U; i < config->banks_len; i++) {
+		sdram.State = HAL_SDRAM_STATE_RESET;
+		memcpy(&sdram.Init, &config->banks[i].init, sizeof(sdram.Init));
+
+		(void)HAL_SDRAM_Init(
+			&sdram,
+			(FMC_SDRAM_TimingTypeDef *)&config->banks[i].timing);
+	}
+
+	/* SDRAM initialization sequence */
+	if (config->banks_len == 2U) {
+		sdram_cmd.CommandTarget = FMC_SDRAM_CMD_TARGET_BANK1_2;
+	} else if (config->banks[0].init.SDBank == FMC_SDRAM_BANK1) {
+		sdram_cmd.CommandTarget = FMC_SDRAM_CMD_TARGET_BANK1;
+	} else {
+		sdram_cmd.CommandTarget = FMC_SDRAM_CMD_TARGET_BANK2;
+	}
+
+	sdram_cmd.AutoRefreshNumber = config->num_auto_refresh;
+	sdram_cmd.ModeRegisterDefinition = config->mode_register;
+
+	/* enable clock */
+	sdram_cmd.CommandMode = FMC_SDRAM_CMD_CLK_ENABLE;
+	(void)HAL_SDRAM_SendCommand(&sdram, &sdram_cmd, 0U);
```

Comment:
I think SDRAM HAL API design is a bit confusing. SDRAM is a single controller that is able to manage 2 banks. In theory, you should have 2 instances of `SDRAM_HandleTypeDef` both pointing to the same controller. However, when you have 2 banks, commands need to be sent just once (setting target FMC_SDRAM_CMD_TARGET_BANK1_2), but the state will just get updated on the instance used to send the commands. It's not relevant here, but in my opinion the API is not as clear as LL counterpart. In order to keep things simple, I decided to use a single `SDRAM_HandleTypeDef` and reset its state when initializing each bank.